File: app/src/main/java/com/example/speedtest_rework/common/NetworkUtils.java
package com.example.speedtest_rework.common;

import android.annotation.TargetApi;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkInfo;
import android.net.wifi.ScanResult;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;

import java.math.BigInteger;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteOrder;
import java.util.List;

public class NetworkUtils {
    public static boolean isConnected(@NonNull Context context) {
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo netInfo = cm.getActiveNetworkInfo();
        if (netInfo != null) {
            NetworkInfo networkInfo = cm.getActiveNetworkInfo();
            return networkInfo != null && networkInfo.isConnected();
        }
        return false;
    }

    public static boolean isWifiConnected(@NonNull Context context) {
        return isConnected(context, ConnectivityManager.TYPE_WIFI);
    }

    public static boolean isWifiConnected(@NonNull Context context, String wifi_name) {
        if (wifi_name.length() > 0 && isWifiConnected(context) && getNameWifi(context) == wifi_name) {
            Log.d("TAG", "isWifiConnected: " + wifi_name);
            return true;
        }
        return false;
    }

    public static boolean isWifiEnabled(@NonNull Context context) {
        WifiManager cm = (WifiManager) context.getApplicationContext().getSystemService(context.WIFI_SERVICE);
        return cm == null && cm.isWifiEnabled();
    }

    public static boolean isMobileConnected(@NonNull Context context) {
        return isConnected(context, ConnectivityManager.TYPE_MOBILE);
    }

    public static NetworkInfo getInforMobileConnected(@NonNull Context context) {
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(context.CONNECTIVITY_SERVICE);
        if (cm != null) {
            NetworkInfo networkInfo = cm.getActiveNetworkInfo();
            return networkInfo;
        }
        return null;


    }

    private static boolean isConnected(@NonNull Context context, int type) {
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(context.CONNECTIVITY_SERVICE);
        NetworkInfo netInfo = cm.getActiveNetworkInfo();

        if (netInfo != null) {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
                NetworkInfo networkInfo = cm.getNetworkInfo(type);
                return networkInfo != null && networkInfo.isConnected();
            } else {
                return isConnected(cm, type);
            }
        }
        return false;

    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    private static boolean isConnected(@NonNull ConnectivityManager cm, int type) {
        Network[] networks = cm.getAllNetworks();
        NetworkInfo networkInfo;
        for (Network network : networks) {
            networkInfo = cm.getNetworkInfo(network);
            if (networkInfo != null && networkInfo.getType() == type && networkInfo.isConnected()) {
                return true;
            }
        }
        return false;
    }

    public static int getDownloadSpeed(@NonNull Context context) {
        int downloadSpeed = 0;
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(context.CONNECTIVITY_SERVICE);
        if (isWifiConnected(context)) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                NetworkCapabilities nc = cm.getNetworkCapabilities(cm.getActiveNetwork());
                downloadSpeed = nc.getLinkDownstreamBandwidthKbps();
                return downloadSpeed;
            }
        }
        return downloadSpeed;
    }

    @TargetApi(Build.VERSION_CODES.M)
    public static int getUploadSpeed(@NonNull Context context) {
        int uploadSpeed = 0;
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(context.CONNECTIVITY_SERVICE);

        if (isWifiConnected(context)) {
            NetworkCapabilities nc = cm.getNetworkCapabilities(cm.getActiveNetwork());
            uploadSpeed = nc.getLinkUpstreamBandwidthKbps();
            return uploadSpeed;
        }
        return uploadSpeed;
    }

    public static WifiInfo getWifiInfo(@NonNull Context context) {
        WifiManager cm = (WifiManager) context.getApplicationContext().getSystemService(context.WIFI_SERVICE);
        return cm.getConnectionInfo();
    }
//    public static String getNameWifi(@NonNull Context context) {
//        ConnectivityManager connManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
//        NetworkInfo networkInfo = connManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
//        if (networkInfo.isConnected()) {
//            WifiManager wifiManager = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
//            WifiInfo wifiInfo = wifiManager.getConnectionInfo();
//            return wifiInfo.toString();
//        }
//        return null;
//    }

    public static String getNameWifi(Context context) {
        WifiManager manager = (WifiManager) context.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        if (manager != null && manager.isWifiEnabled()) {
            WifiInfo wifiInfo = manager.getConnectionInfo();
            if (wifiInfo != null) {
                NetworkInfo.DetailedState state = WifiInfo.getDetailedStateOf(wifiInfo.getSupplicantState());
                if (state == NetworkInfo.DetailedState.CONNECTED || state == NetworkInfo.DetailedState.OBTAINING_IPADDR) {
                    return wifiInfo.getSSID().substring(1, wifiInfo.getSSID().length() - 1);
                }
            }
        }
        return null;
    }

    public static String getNameWifi(WifiManager manager) {
        if (manager != null && manager.isWifiEnabled()) {
            WifiInfo wifiInfo = manager.getConnectionInfo();
            if (wifiInfo != null) {
                NetworkInfo.DetailedState state = WifiInfo.getDetailedStateOf(wifiInfo.getSupplicantState());
                if (state == NetworkInfo.DetailedState.CONNECTED || state == NetworkInfo.DetailedState.OBTAINING_IPADDR) {
                    return wifiInfo.getSSID().substring(1, wifiInfo.getSSID().length() - 1);
                }
            }
        }
        return null;
    }

    public static List<ScanResult> getListWifi(Context context, WifiManager manager) {

        if (manager != null && manager.isWifiEnabled()) {
            List<ScanResult> wifiList = manager.getScanResults();
            return wifiList;
        }
        return null;
    }

    public static String wifiIpAddress(Context context) {
        WifiManager wifiManager = (WifiManager) context.getSystemService(Context.WIFI_SERVICE);
        int ipAddress = wifiManager.getConnectionInfo().getIpAddress();

        // Convert little-endian to big-endianif needed
        if (ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)) {
            ipAddress = Integer.reverseBytes(ipAddress);
        }

        byte[] ipByteArray = BigInteger.valueOf(ipAddress).toByteArray();

        String ipAddressString;
        try {
            ipAddressString = InetAddress.getByAddress(ipByteArray).getHostAddress();
        } catch (UnknownHostException ex) {
            Log.e("WIFIIP", "Unable to get host address.");
            ipAddressString = null;
        }

        return ipAddressString;
    }

    public static int[] getRangeWifi(int channel, int type) {
        int r_range = 0;
        int l_range = 0;
        switch (type) {
            case 0:
                if (channel <= 2) {
                    r_range = 0;
                    l_range = 4;
                } else {
                    r_range = channel - 2;
                    l_range = channel + 2;
                }
                break;

            case 1:
                if (channel - 4 > 0) {
                    r_range = channel - 2 > 0 ? channel - 2 : 0;
                    l_range = r_range + 8;
                } else if (channel + 4 > 14) {
                    l_range = channel + 2 < 14 ? channel + 2 : 14;
                    r_range = l_range - 8;
                }
                break;

        }

        int result[] = {r_range, l_range};
        return result;
    }

    public static int convertFreqtoChannel(int freq) {
        if (freq == 2484)
            return 14;

        if (freq < 2484)
            return (freq - 2407) / 5;

        return freq / 5 - 1000;
    }

    public static boolean isNetworkConnected(Context context) {
        ConnectivityManager connectivityManager =
                (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);

        // For 29 api or above
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(connectivityManager.getActiveNetwork());
            if (capabilities == null) {
                return false;
            }
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                return true;
            } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
                return true;
            } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
                return true;
            } else {
                return false;
            }
        } else {
            if (connectivityManager.getActiveNetworkInfo() != null && connectivityManager.getActiveNetworkInfo().isConnectedOrConnecting()) {
                return true;
            } else {
                return false;
            }
        }
    }


}

File: app/src/main/java/com/example/speedtest_rework/common/NetworkUtils.kt
package com.example.speedtest_rework.common

import android.annotation.TargetApi
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.net.NetworkInfo
import android.net.wifi.ScanResult
import android.net.wifi.WifiInfo
import android.net.wifi.WifiManager
import android.os.Build
import android.util.Log
import java.math.BigInteger
import java.net.InetAddress
import java.net.UnknownHostException
import java.nio.ByteOrder

 object NetworkUtils {

         fun isConnected(context: Context): Boolean {
             val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
             val netInfo = cm.activeNetworkInfo
             if (netInfo != null) {
                 val networkInfo = cm.activeNetworkInfo
                 return networkInfo != null && networkInfo.isConnected
             }
             return false
         }
         @JvmStatic
         fun isWifiConnected(context: Context): Boolean {
             return isConnected(context, ConnectivityManager.TYPE_WIFI)
         }

         fun isWifiConnected(context: Context, wifi_name: String): Boolean {
             if (wifi_name.isNotEmpty() && isWifiConnected(context) && getNameWifi(context) === wifi_name) {
                 Log.d("TAG", "isWifiConnected: $wifi_name")
                 return true
             }
             return false
         }

         fun isWifiEnabled(context: Context): Boolean {
             val cm = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
             return cm == null && cm.isWifiEnabled
         }

         fun isMobileConnected(context: Context): Boolean {
             return isConnected(context, ConnectivityManager.TYPE_MOBILE)
         }

         fun getInforMobileConnected(context: Context): NetworkInfo? {
             val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
             return if (cm != null) {
                 cm.activeNetworkInfo
             } else null
         }

         private fun isConnected(context: Context, type: Int): Boolean {
             val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
             val netInfo = cm.activeNetworkInfo
             return if (netInfo != null) {
                 if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
                     val networkInfo = cm.getNetworkInfo(type)
                     networkInfo != null && networkInfo.isConnected
                 } else {
                     isConnected(cm, type)
                 }
             } else false
         }

         @TargetApi(Build.VERSION_CODES.LOLLIPOP)
         private fun isConnected(cm: ConnectivityManager, type: Int): Boolean {
             val networks = cm.allNetworks
             var networkInfo: NetworkInfo?
             for (network in networks) {
                 networkInfo = cm.getNetworkInfo(network)
                 if (networkInfo != null && networkInfo.type == type && networkInfo.isConnected) {
                     return true
                 }
             }
             return false
         }

         fun getDownloadSpeed(context: Context): Int {
             var downloadSpeed = 0
             val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
             if (isWifiConnected(context)) {
                 if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                     val nc = cm.getNetworkCapabilities(cm.activeNetwork)
                     downloadSpeed = nc!!.linkDownstreamBandwidthKbps
                     return downloadSpeed
                 }
             }
             return downloadSpeed
         }

         @TargetApi(Build.VERSION_CODES.M)
         fun getUploadSpeed(context: Context): Int {
             var uploadSpeed = 0
             val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
             if (isWifiConnected(context)) {
                 val nc = cm.getNetworkCapabilities(cm.activeNetwork)
                 uploadSpeed = nc!!.linkUpstreamBandwidthKbps
                 return uploadSpeed
             }
             return uploadSpeed
         }

         fun getWifiInfo(context: Context): WifiInfo? {
             val cm = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
             return cm.connectionInfo
         }

         fun getNameWifi(context: Context): String? {
             val manager =
                 context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
             if (manager != null && manager.isWifiEnabled) {
                 val wifiInfo = manager.connectionInfo
                 if (wifiInfo != null) {
                     val state = WifiInfo.getDetailedStateOf(wifiInfo.supplicantState)
                     if (state == NetworkInfo.DetailedState.CONNECTED || state == NetworkInfo.DetailedState.OBTAINING_IPADDR) {
                         return wifiInfo.ssid.substring(1, wifiInfo.ssid.length - 1)
                     }
                 }
             }
             return null
         }

         fun getNameWifi(manager: WifiManager?): String? {
             if (manager != null && manager.isWifiEnabled) {
                 val wifiInfo = manager.connectionInfo
                 if (wifiInfo != null) {
                     val state = WifiInfo.getDetailedStateOf(wifiInfo.supplicantState)
                     if (state == NetworkInfo.DetailedState.CONNECTED || state == NetworkInfo.DetailedState.OBTAINING_IPADDR) {
                         return wifiInfo.ssid.substring(1, wifiInfo.ssid.length - 1)
                     }
                 }
             }
             return null
         }

         fun getListWifi(context: Context?, manager: WifiManager?): List<ScanResult?>? {
             return if (manager != null && manager.isWifiEnabled) {
                 manager.scanResults
             } else null
         }

         fun wifiIpAddress(context: Context): String? {
             val wifiManager = context.getSystemService(Context.WIFI_SERVICE) as WifiManager
             var ipAddress = wifiManager.connectionInfo.ipAddress

             // Convert little-endian to big-endianif needed
             if (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN) {
                 ipAddress = Integer.reverseBytes(ipAddress)
             }
             val ipByteArray = BigInteger.valueOf(ipAddress.toLong()).toByteArray()
             val ipAddressString: String?
             ipAddressString = try {
                 InetAddress.getByAddress(ipByteArray).hostAddress
             } catch (ex: UnknownHostException) {
                 null
             }
             return ipAddressString
         }

         fun getRangeWifi(channel: Int, type: Int): IntArray? {
             var r_range = 0
             var l_range = 0
             when (type) {
                 0 -> if (channel <= 2) {
                     r_range = 0
                     l_range = 4
                 } else {
                     r_range = channel - 2
                     l_range = channel + 2
                 }
                 1 -> if (channel - 4 > 0) {
                     r_range = if (channel - 2 > 0) channel - 2 else 0
                     l_range = r_range + 8
                 } else if (channel + 4 > 14) {
                     l_range = if (channel + 2 < 14) channel + 2 else 14
                     r_range = l_range - 8
                 }
             }
             return intArrayOf(r_range, l_range)
         }

         fun convertFreqtoChannel(freq: Int): Int {
             if (freq == 2484) return 14
             return if (freq < 2484) (freq - 2407) / 5 else freq / 5 - 1000
         }

         fun isNetworkConnected(context: Context): Boolean {
             val connectivityManager =
                 context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

             // For 29 api or above
             return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                 val capabilities =
                     connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
                         ?: return false
                 if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                     true
                 } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
                     true
                 } else capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)
             } else {
                 connectivityManager.activeNetworkInfo != null && connectivityManager.activeNetworkInfo!!
                     .isConnectedOrConnecting
             }
         }

}